worker/src/socket.rs

```diff
@@ -0,0 +1,340 @@
+use std::task::Poll;
+
+use crate::r2::js_object;
+use crate::Result;
+use futures_util::FutureExt;
+use js_sys::{Boolean as JsBoolean, JsString, Number as JsNumber, Object as JsObject, Uint8Array};
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::JsFuture;
+use worker_sys::console_log;
+
+enum Reading {
+    None,
+    Pending(JsFuture),
+    Ready(Vec<u8>),
+}
+
+impl Default for Reading {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Writing {
+    Pending(JsFuture, usize),
+    None,
+}
+
+impl Default for Writing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Closing {
+    Pending(JsFuture),
+    None,
+}
+
+impl Default for Closing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// Represents an outbound TCP connection from your Worker.
+pub struct Socket {
+    inner: worker_sys::Socket,
+    // TODO: Box this stuff so it can be sync / send?
+    _writable: worker_sys::WritableStream,
+    writer: worker_sys::WritableStreamDefaultWriter,
+    _readable: worker_sys::ReadableStream,
+    reader: worker_sys::ReadableStreamDefaultReader,
+    write: Option<Writing>,
+    read: Option<Reading>,
+    close: Option<Closing>,
+}
+
+impl Socket {
+    fn new(inner: worker_sys::Socket) -> Self {
+        let _writable = inner.writable();
+        let writer = _writable.get_writer();
+        let _readable = inner.readable();
+        let reader = _readable.get_reader();
+        Socket {
+            inner,
+            _writable,
+            writer,
+            _readable,
+            reader,
+            read: None,
+            write: None,
+            close: None,
+        }
+    }
+
+    /// Closes the TCP socket. Both the readable and writable streams are forcibly closed.
+    pub async fn close(&mut self) -> Result<()> {
+        JsFuture::from(self.inner.close()).await?;
+        Ok(())
+    }
+
+    /// This Future is resolved when the socket is closed
+    /// and is rejected if the socket encounters an error.
+    pub async fn closed(&self) -> Result<()> {
+        JsFuture::from(self.inner.closed()).await?;
+        Ok(())
+    }
+
+    /// Upgrades an insecure socket to a secure one that uses TLS,
+    /// returning a new Socket. Note that in order to call this method,
+    /// you must set [`secure_transport`](SocketOptions::secure_transport)
+    /// to [`StartTls`](SecureTransport::StartTls) when initially
+    /// calling [`connect`](connect) to create the socket.
+    pub async fn start_tls(self) -> Socket {
```

Comment:
Should this be an `async` function? the JS API doesn't return a Promise here

Reply:
Does it? Our docs do not indicate that. https://developers.cloudflare.com/workers/runtime-apis/tcp-sockets/#socket

I have not tested this.

Reply:
Hm, am I missing something? The docs don't show it returning a Promise:
<img width="856" alt="Screenshot 2023-05-24 at 19 24 48" src="https://github.com/cloudflare/workers-rs/assets/246651/37273c76-b664-43a8-bf42-804568b9ec82">

Reply:
Oh, I see what you are saying, I thought you meant that it should be async, yep, will fix.